### DISCUSSION TITLE
Cache invalidation/Pagination/Search

### DISCUSSION THREAD

**Discussion by Miloshinjo** [urql-graphql/urql#1612] [Q&A]
Hello,

I have a particular problem, where I am using Elastic Search raw and they have their own pagination so I am writing a custom pagination function and all of that is going fine. I made pagination work, it's working fine etc... 

Where I am having an issue is as follows: 

When I make a new search , I want to invalidate all prior searches in the cache  (because if I don't it's causing me a lot of problems with double results etc, my pagination function is not perfect by any means). Is it somehow possible to invalidate all prior queries from the cache in a `resolver` function (because I tried that and somehow it doesn't work). Or can we invalidate it when calling the query or something like that? All I found in docs is invalidation during mutations. 

p.s. I don't wanna provide code samples cause it's not a personal project, but I will make some pseudo-code that sums up my situation if needed :) 

Big Thanks.

**Comment by JoviDeCroock** [comment]
Hey,

"invalidating" should be automatically done if you follow the recipe used [here](https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/extras/simplePagination.ts) the [`compareArgs`](https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/extras/simplePagination.ts#L17) method will ensure you are only paginating within one filter/sort segment.

**Reply by Miloshinjo** [reply]
Hi, thanks for a quick reply! :) 

I was following that function in making my own and I'll take a better look where I made the mistake. The thing is that `arguments` always change in Elastic search since it has `size` and `from` that are part of the `query` which makes `arguments` change all the time. My `compareArgs` tries to isolate only the `search` part of the arguments and compare that. The main problem lies that some results repeat from search to search if I look for similar stuff and cause cache is normalized, searchResult won't repeat in cache so if I exclude something that was fetched in the previous search with previous arguments, that won't show up in the new search. 

But I will look even deeper and find the mistake, it's 99% that I made an error, just wanted to maybe see if there is a solution here. 

Thanks again.

**Reply by JoviDeCroock** [reply]
It's hard for us to visualize your exact use case as we're not seeing how your ES pagination works 😅 last time I used ES with search-pagination I remember that basically it's the `simplePagination`, `from` is the same in ES but `limit` becomes `size`.

`search` is your segment (or filter) and the only thing that needs to be excluded from comparing are `limit` and `size`.

**Reply by Miloshinjo** [reply]
I know, I am sorry :) but yes, it's very same as simplePagination, just with one difference, arguments are passed as a string. 

```  
const [{ data, fetching, error }] = useSearchQuery({
    variables: {
      query: "{"query":{"query_string":{"query":"message 1"}},"from":0,"size":10}",
    },
  });
```

As you can see, it's in form of a string, which I parse and I have my own compareArgs which does something like 

```
const compareArgs = (
  fieldArgs: Variables,
  connectionArgs: Variables | null,
): boolean => {
  if (
    connectionArgs === null ||
    typeof fieldArgs.query !== 'string' ||
    typeof connectionArgs.query !== 'string'
  ) {
    return false;
  }

  const { query: fieldQuery } = JSON.parse(fieldArgs.query);
  const { query: connectionQuery } = JSON.parse(connectionArgs.query);

  return JSON.stringify(fieldQuery) === JSON.stringify(connectionQuery);
};
```

**Reply by Miloshinjo** [reply]
Again, if you don't see nothing , I will dig more, just fished for a solution or some technique that some might share, don't beat your head with this :) I thank you for looking at it at all

**Reply by JoviDeCroock** [reply]
I mean, if you're not connecting to ES directly you can just parse this normally as an object and do the transformation step from POJO --> ES string in your GraphQL layer as this currently seems to leak your backing DB's implementation details in the FrontEnd 😅 

In case you don't have control over that you'll probably have to write a compare that takes the shape of your query string in account and does the translation steps itself. This is a server-side detail normally though, you can also opt to not do it in a resolver and just use `useState` & `useEffect` as that will give you more granular control over when to reset your list.

**Reply by Miloshinjo** [reply]
Alright, thanks for the help a lot.

One last thing just to be clear, invalidating cache from the resolver function is a no-go, cause of async stuff etc..? (I gave it a half-assed try but did not work).

**Reply by JoviDeCroock** [reply]
GraphCache attempts to be a client-side representation of your server-side state, this means that we only allow to perform mutations on your cache as part of new data coming in. This can only happen during `mutations` or `subscriptions` hence why we don't allow data-mutations during retrieval phases. This is similar to how the GraphQL pushes for queries to be side-effect free.

**Reply by Miloshinjo** [reply]
Thank you for that, it steers me away from a very bad course :) 

And thanks for taking the time to chat about this.

**Reply by Miloshinjo** [reply]
Hi, I just wanna let you know that I resolved the situation and that the problem was really silly and that urql does it's job very nice and my compareArgs is fine too :) I am too ashamed to say what made it not work, but I can promise you it's not a semicolon :) 

thanks again for this amazing library.
